### ISSUE TITLE
Consider wrapping Patch2Self DWI denoising technique into SCT

### ISSUE THREAD

**Issue by jcohenadad** [spinalcordtoolbox/spinalcordtoolbox#3234]
The recent patch-based DWI denoising technique from Fadnavis et al. seems promising 👉   [article](https://papers.nips.cc/paper/2020/file/bc047286b224b7bfa73d4cb02de1238d-Paper.pdf) | [tutorial](https://www.youtube.com/watch?v=S4MoCox98M8&feature=emb_title&ab_channel=DiffusionImaginginPython).

We should test it and consider adding an SCT wrapper for DWI pipeline.

**Comment by ShreyasFadnavis** [comment]
I can help with this! Is there a dataset within SCT that I can try it on?

**Comment by jcohenadad** [comment]
> I can help with this! Is there a dataset within SCT that I can try it on?

Awesome! If you could use our [testing data](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming%3A-Test-data#real-anatomical-data-sct_testing_data) that would be ideal, so that your code could seamlessly be integrated into the testing suite.

**Comment by ShreyasFadnavis** [comment]
Hi @jcohenadad ! I tried denoised [this](https://github.com/sct-data/sct_testing_data/blob/master/dmri/dmri.nii.gz) data. Here is the output that one would get from `Patch2Self`. 

![image](https://user-images.githubusercontent.com/17598557/109037313-96ada100-7698-11eb-877a-7220bad78e1f.png)

I noticed that this dataset is too small. But as the size gets bigger, the output will get better. Here are the details of running the algorithm:
```
import numpy as np

from dipy.io.image import load_nifti, save_nifti
from dipy.denoise.patch2self import patch2self

bvals = np.array([0, 750, 750, 750, 750, 750, 750])
data, affine = load_nifti('dmri.nii.gz')

# no parameter tuning needed 
denoised_arr  = patch2self(data, bvals, model='ols', verbose=True)
```
### Total time taken for Patch2Self:  0.07434701919555664  seconds

Does this look okay? Is there some other dataset I should run this on?

**Comment by ShreyasFadnavis** [comment]
Ping @jcohenadad :)

**Comment by joshuacwnewton** [comment]
> We should test it and consider adding an SCT wrapper for DWI pipeline.

One quick observation: SCT currently has a tool called `sct_denoising_onlm`, which is a wrapper for `dipy.denoise.nlmeans`. Because both are `dipy` denoising methods, I imagine there would be a lot of code overlap between the current `sct_denoising_onlm` and a hypothetical `sct_denoising_patch2self`. 

This is another one of those cases where sub-commands would be useful, as mentioned by @Drulex in https://github.com/neuropoly/spinalcordtoolbox/issues/3159#issuecomment-768316706. Sub-commands would allow us to do something like:

* `sct_denoise onlm <onlm arguments>`
* `sct_denoise patch2self <patch2self arguments>`

That way we group functionality but still have separate parsers.

**Comment by jcohenadad** [comment]
> Does this look okay? 

looks awesome!

> Is there some other dataset I should run this on?

sct_example_data -- or https://github.com/spine-generic/data-multi-subject#spine-generic-public-database-multi-subject

> Ping @jcohenadad :)

sorry-- i did get the notification but time is my enemy here...

> One quick observation: SCT currently has a tool called sct_denoising_onlm, which is a wrapper for dipy.denoise.nlmeans. Because both are dipy denoising methods, I imagine there would be a lot of code overlap between the current sct_denoising_onlm and a hypothetical sct_denoising_patch2self.

absolutely-- since this function is dmri-specific, we definitely want to name it with prefix sct_dmri_
now question is: do we want to keep denoising_onlm? should we merge all methods into `sct_dmri_denoise` and have a flag to select the method?

**Comment by ShreyasFadnavis** [comment]
Thank you @jcohenadad -- Sorry for bothering 👍🏽  

For the example data - will have it denoised shortly!
I think we can keep both `denoising_onlm` and `patch2self` -- perhaps set `patch2self` as default?
@Garyfallidis what do you think? Should we replace `denoising_onlm` completely?

**Comment by joshuacwnewton** [comment]
Slight correction: I've just noticed that SCT is currently wrapping `dipy.denoise.nlmeans` in two separate places! :fearful: 
1. `sct_maths -denoise`,  which simply applies the filter.
2. `sct_denoising_onlm`, which applies the filter _and_ generates a difference image, and can also generates plots if requested.

This is a separate from the question of "`nlmeans` vs `patch2self`", but some cleanup might be needed... :sweat_smile:

**Comment by Garyfallidis** [comment]
My 2 cents here. For DWI data be happy to replace nlmeans with patch2self. For T1 data (single volume data) you may want to keep NLMeans.

**Comment by Garyfallidis** [comment]
Replacing meaning here change default option or remove if you want to reduce number of options in your interface.

**Comment by jcohenadad** [comment]
> Where should denoising code live?

I have a slight preference for `sct_denoise`, because denoising concerns both dmri and non-dmri scans (so we shouldn't wrap this under `sct_dmri_` CLI).

**Comment by ShreyasFadnavis** [comment]
> > Where should denoising code live?
> 
> I have a slight preference for `sct_denoise`, because denoising concerns both dmri and non-dmri scans (so we shouldn't wrap this under `sct_dmri_` CLI).

Hi @jcohenadad! Sounds good. On a similar note `Patch2Self` should work for fMRI too :) Is there some dataset I can test that on too? I understand this is more involved than dMRI, where the analysis needs to be done on a cohort!

**Comment by jcohenadad** [comment]
> On a similar note Patch2Self should work for fMRI too :)

awesome! 🎉   i know many people who would be happy about that

> is there some dataset I can test that on too?

sct_testing_data/fmri

thanks!!

**Comment by ShreyasFadnavis** [comment]
Hi @joshuacwnewton I am willing to help with this 👍🏽 What would a PR for making this interface entail? I guess you are waiting on some prior PR to be merged 1st?

**Comment by joshuacwnewton** [comment]
> Hi @joshuacwnewton I am willing to help with this 👍🏽 

Thank you so much for offering to contribute! This would be much appreciated. :slightly_smiling_face: 

> I guess you are waiting on some prior PR to be merged 1st?

Yes. Some high-priority bugs have been discovered in the recently-released SCT v5.2.0. So, it's possible that a patch release will take priority over new features (at least in the very short-term). I've only just brought this up with the other SCT developers, though, so this is just me speaking personally, but I thought I'd give you a heads-up.

> What would a PR for making this interface entail? 

I think I agree with @jcohenadad's opinion in https://github.com/neuropoly/spinalcordtoolbox/issues/3234#issuecomment-786199429. In other words, the PR would update the existing `sct_denoising_onlm` script to become `sct_denoise`, and would aim to support both of the dipy denoising methods.

For now, then, I think the best place to start would be to take a look at the [existing script](https://github.com/neuropoly/spinalcordtoolbox/blob/master/spinalcordtoolbox/scripts/sct_denoising_onlm.py). It's fairly isolated from the rest of the SCT package, so you should be able to focus on just that file, without having to worry too much about the other modules in SCT. Some things to possibly think about:

1. The current structure of the script.
    * i.e. How it's organized, which parts are broken up into functions, etc.
    * There are some relevant notes about this in our developer Wiki, on the page [**CLI script conventions**](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming:-CLI-script-conventions-(structure,-argparse)).
2. The existing user interface (`argparse`)
    * How could the current setup be adapted to support both denoising methods?
    * How would the user choose one method or the other, then provide input arguments for that chosen method? 
    * Which input arguments overlap between methods, and which are unique?
    * Perhaps `argparse`'s [subparsers](https://docs.python.org/3/library/argparse.html#sub-commands) might be of value here.

(One last note for context: The original script was written in 2015, and AFAIK SCT's development processes have changed quite a bit since then. So, it's possible that there's room for improvement in the style/structure of the script, too.)

**Comment by ShreyasFadnavis** [comment]
Hi @joshuacwnewton! Gotcha! Thank you for the detailed response... I think its very less work 💯 I took a look at the script and both `Patch2Self ` + `ONLM` have the exact same interface in DIPY -- basically some copy-pasting 👍🏽

**Comment by joshuacwnewton** [comment]
That sounds good! Thank you again for taking a look at this. :smile:

I should be available as a reviewer, so feel free to tag me in the PR, as well.

**Comment by joshuacwnewton** [comment]
Ah! This was addressed by #3510, but the issue was never linked. Closing now. :)
